internal/pkg/composable/providers/kubernetesleaderelection/kubernetes_leaderelection_test.go

```diff
@@ -0,0 +1,218 @@
+		comm := ctesting.NewContextComm(ctx)
+
+		os.Setenv("POD_NAME", podNames[i])
+		go p.Run(ctx, comm)
+
+		if i == 1 {
+			break
+		}
+
+		// We need to wait for the first agent to acquire the lease, so we can POD_NAME environment variable again
+		expectedLeader := leaderElectorPrefix + podNames[i]
+		for {
+			holder, err := getLeaseHolder(client)
+			require.NoError(t, err)
+
+			if holder == expectedLeader {
+				break
+			}
+		}
+	}
+
+	done := make(chan int)
+	// At this point the current holder is agent1. Let's change it to agent2.
+	// We do this as a goroutine, so we can have a timeout to avoid this function running forever.
+	go func() {
+		for {
+			// Force the lease to be applied again, so a new leader is elected.
+			intermediateHolder := "does-not-matter"
+			lease.Spec.HolderIdentity = &intermediateHolder
+			err = applyLease(client, lease, false)
+			require.NoError(t, err)
+
+			// Make sure the holder changed
+			currentHolder, err := getLeaseHolder(client)
+			require.NoError(t, err)
+			require.Equal(t, intermediateHolder, currentHolder)
+
+			for {
+				currentHolder, err = getLeaseHolder(client)
+				require.NoError(t, err)
+
+				// In this case, we already have an agent as holder
+				if currentHolder != intermediateHolder {
+					break
+				}
+			}
+
+			if currentHolder == leaderElectorPrefix+podNames[1] {
+				done <- 1
+				break
+			}
+		}
+	}()
+
+	select {
+	case <-done:
+	case <-time.After(time.Duration(leaseDuration+leaseRetryPeriod) * 20 * time.Second):
+		require.FailNow(t, "Waited for 20 lease durations, but agent2 still not acquired the lease. This error "+
+			"is possible, but unlikely.")
+	}
+
+	// Now that the holder is agent2, let's wait for agent1 to be reelected.
+	// To avoid having to wait very long, the context of agent2 will be canceled so the leader elector will not be
+	// running anymore. This way there is only one instance fighting to acquire the lease.
+	cancelFuncs[1]()
+	// We do this as a goroutine, so we can have a timeout to avoid this function running forever.
+	go func() {
+		for {
+			// Force the lease to be applied again, so a new leader is elected.
+			intermediateHolder := "does-not-matter"
+			lease.Spec.HolderIdentity = &intermediateHolder
+			err = applyLease(client, lease, false)
+			require.NoError(t, err)
+
+			// Make sure the holder changed
+			var currentHolder string
+			for {
+				currentHolder, err = getLeaseHolder(client)
+				require.NoError(t, err)
+				if currentHolder == intermediateHolder {
+					break
+				}
+			}
+
+			for {
+				currentHolder, err = getLeaseHolder(client)
+				require.NoError(t, err)
+
+				// In this case, we already have an agent as holder
+				if currentHolder != intermediateHolder {
+					break
+				}
+			}
+
+			if currentHolder == leaderElectorPrefix+podNames[0] {
+				done <- 1
+				break
+			}
+		}
+	}()
+
+	select {
+	case <-done:
+	case <-time.After(time.Duration(leaseDuration+leaseRetryPeriod) * 3 * time.Second):
```

Comment:
I also added a timeout here to avoid waiting this test running for a long, long time. I cannot think of reasons why this would ever fail here, since we give a long time for the lease to be acquired.